lib/cousins.ml

```diff
@@ -73,3 +73,302 @@ let br_inter_is_empty b1 b2 =
+  in
+  let max_l1, max_l2 = max_l1_l2_aux max_a_l max_d_l in
+  let max_i =
+    let rec loop0 i =
+      if cousins_cnt.(i).(0) <> [] && i < max_l1 then loop0 (i + 1) else i
+    in
+    loop0 0
+  in
+  let rec loop i j =
+    if cousins_cnt.(i).(j) <> [] then
+      if j < max_l2 then loop i (j + 1) else (max_i, j - i)
+    else if i < max_l1 && j < max_l2 then loop (i + 1) (j + 1)
+    else (max_i, j - i)
+  in
+  loop 0 0
+
+let cousins_l1_l2_aux base max_a_l max_d_l l1 l2 p =
+  let il1 = int_of_string l1 in
+  let il2 = int_of_string l2 in
+  let max_l1, max_l2 = max_l1_l2_aux max_a_l max_d_l in
+  if il1 <= max_l1 && il2 - il1 <= max_l2 then
+    let cousins_cnt, _cousins_dates = init_cousins_cnt base max_a_l max_d_l p in
+    (* gros calcul *)
+    Some cousins_cnt.(il1).(il2)
+  else None
+
+(* create a new l (ip, (ifaml, iancl, cnt), lev) from (ip, ifaml, ipar, lev) *)
+let cousins_fold l =
+  let same_ifaml ifl1 ifl2 =
+    List.for_all2 (fun if1 if2 -> if1 = if2) ifl1 ifl2
+  in
+  let l = List.sort compare l in
+  let rec loop first acc (ip0, (ifaml0, iancl0, cnt0), lev0) = function
+    | (ip, ifaml, ianc, lev) :: l when ip = ip0 && same_ifaml ifaml ifaml0 ->
+        loop false acc
+          ( ip,
+            ( ifaml,
+              (if List.mem ianc iancl0 then iancl0 else ianc :: iancl0),
+              cnt0 + 1 ),
+            lev @ lev0 )
+          l
+    | (ip, ifaml, ianc, lev) :: l ->
+        loop false
+          (if first || cnt0 = 0 then acc
+          else (ip0, (ifaml0, iancl0, cnt0), lev0) :: acc)
+          (ip, (ifaml, [ ianc ], 1), lev)
+          l
+    | [] ->
+        if first || cnt0 = 0 then acc
+        else (ip0, (ifaml0, iancl0, cnt0), lev0) :: acc
+  in
+  loop false [] (Gwdb.dummy_iper, ([], [], 0), [ 0 ]) l
+
+let asc_cnt_t = ref None
+let desc_cnt_t = ref None
+
+(* tableau des ascendants de p *)
+let init_asc_cnt base max_a_l p =
+  match !asc_cnt_t with
+  | Some t -> t
+  | None ->
+      let t' =
+        let asc_cnt = Array.make (max_a_l + 1) [] in
+        asc_cnt.(0) <-
+          [ (get_iper p, [ Gwdb.dummy_ifam ], Gwdb.dummy_iper, [ 0 ]) ];
+        for i = 1 to max_a_l do
+          asc_cnt.(i) <- ascendants base [] asc_cnt.(i - 1) i
+        done;
+        asc_cnt
+      in
+      asc_cnt_t := Some t';
+      t'
+
+(* tableau des ascendants de p *)
+let init_desc_cnt base max_d_l p =
+  match !desc_cnt_t with
+  | Some t -> t
+  | None ->
+      let t' =
+        let desc_cnt = Array.make (max_d_l + 1) [] in
+        desc_cnt.(0) <-
+          [ (get_iper p, [ Gwdb.dummy_ifam ], Gwdb.dummy_iper, [ 0 ]) ];
+        for i = 1 to min max_d_l (Array.length desc_cnt - 1) do
+          desc_cnt.(i) <- descendants_aux base desc_cnt.(i - 1) []
+        done;
+        desc_cnt
+      in
+      desc_cnt_t := Some t';
+      t'
+
+let anc_cnt_aux base max_a_l lev at_to p =
+  let asc_cnt =
+    match !asc_cnt_t with Some t -> t | None -> init_asc_cnt base max_a_l p
+  in
```

Comment:
You use `asc_cnt_t` to cache the computation of `init_asc_cnt`.
One way to do it without having to use a ref and an option is to use a closure:

```ocaml
let anc_cnt_aux base max_a_l  p =
   let asc_cnt = init_asc_cnt base max_a_l p in
   fun lev at_to -> 
        ...
```
`init_asc_cnt` will only be called once the first time you call `anc_cnt_aux`.

You can do the same for `desc_cnt_aux`.